### PR TITLE
refactor: disconnect reason [MTT-2984] [MTT-5200]

### PR DESCRIPTION
**Pull request by LPLafontaineB**
### Description

This PR replaces our workaround using custom messages to send a reason to clients when their approval gets denied by the host or when the host shuts down voluntarily with the DisconnectReason feature introduced in NGO 1.2.0. This PR also updates the state diagram in architecture.md to remove the DisconnectingWithReason state that is no longer necessary.

### Issue Number(s)

[MTT-2984](https://jira.unity3d.com/browse/MTT-2984) & [MTT-5200](https://jira.unity3d.com/browse/MTT-5200)

### Contribution checklist
 - [ ] Tests have been added for boss room and/or utilities pack
 - [x] Release notes have been added to the [project changelog](../CHANGELOG.md) file and/or [package changelog](../Packages/com.unity.multiplayer.samples.coop/CHANGELOG.md) file
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] JIRA ticket ID is in the PR title or at least one commit message
 - [x] Include the ticket ID number within the body message of the PR to create a hyperlink
 - [x] An Index entry has been added in readme.md if applicable

### CODE REVIEW

File: Assets/Scripts/ConnectionManagement/ConnectionState/HostingState.cs [hunk 1]
```diff
@@ -71,9 +71,23 @@ public override void OnClientDisconnect(ulong clientId)
 
         public override void OnUserRequestedShutdown()
```
Comment (SamuelBellomo) [inline Assets/Scripts/ConnectionManagement/ConnectionState/HostingState.cs:72]: Should we add this to the index? Something along the lines of "disconnecting all clients with reason", something like that?

File: Assets/Scripts/ConnectionManagement/ConnectionState/HostingState.cs [hunk 2]
```diff
@@ -71,9 +71,23 @@ public override void OnClientDisconnect(ulong clientId)
...
+            {
+                if (id != m_ConnectionManager.NetworkManager.LocalClientId)
+                {
+                    clientIds[i++] = id;
+                }
+            }
+
+            var reason = JsonUtility.ToJson(ConnectStatus.HostEndedSession);
+
+            foreach (var id in clientIds)
```
Comment (SamuelBellomo) [inline Assets/Scripts/ConnectionManagement/ConnectionState/HostingState.cs:86]: why two for loops? Why not just do this in the first loop?

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,10 @@ Additional documentation and release notes are available at [Multiplayer Documen
 ### Changed
 * Bumped RNSM to 1.1.0: Switched x axis units to seconds instead of frames now that it's available. This means adjusting the sample count to a lower value as well to 30 seconds, since the x axis was moving too slowly. (#788)
 * Updated Boss Room to NGO 1.2.0 (#791).
+  * Removed a workaround in our tests waiting for two frames before shutting down a client that is attempting to connect to a server. (see https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/2261)
+
+### Cleanup
```
Comment (SamuelBellomo) [inline CHANGELOG.md:15]: This feels bigger than just a cleanup, this should be at the front of our "changed" or "added" section.

File: Assets/Scripts/ConnectionManagement/ConnectionState/HostingState.cs (hunk 2 above)
Comment (LPLafontaineB) [reply Assets/Scripts/ConnectionManagement/ConnectionState/HostingState.cs:86]: Calling DisconnectClient removes clients from the loop which is not allowed while iterating. I suppose I could replace it with a single for(int i = Count - 1; i >= 0; i++) instead though if you think that it is cleaner

Comment (SamuelBellomo) [reply Assets/Scripts/ConnectionManagement/ConnectionState/HostingState.cs:86]: i-- (not i++ :P )
but yeah, that'd make this a bit easier to understand.

Comment (LPLafontaineB) [reply Assets/Scripts/ConnectionManagement/ConnectionState/HostingState.cs:86]: will do

**Review by SamuelBellomo — APPROVED** [review]

**Pull request merged**